### PR TITLE
More beginner-friendly TCP server example

### PR DESCRIPTION
**Pull request by lbernick**
I wanted to write an example to help guide people new to asynchronous programming, with fewer external dependencies (as some people requested in #66). The example starts with the [single threaded TCP server from the Rust book](https://doc.rust-lang.org/book/ch20-01-single-threaded.html) and uses async functionality to add concurrency.
The example goes through:
- some examples of blocking code and how to make them non blocking (should address #64 as well)
- spawning tasks vs using combinators to run them concurrently on a single thread
- testing code by creating mocks and implementing async read/write traits for the mocks

Happy to receive any feedback.

### CODE REVIEW

File: src/08_example/02_handling_connections_concurrently.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+we can't read a new request from this stream until we're done with the previous one.
+One strategy to work around this is to spawn a new Task to handle each connection in the background:
+```rust
+{{#include ../../examples/08_03_concurrent_tcp_server/src/main.rs:main_func}}
+```
+
+This works because under the hood, the `async_std` executor runs `handle_connection` on a separate thread.
+However, this doesn't completely solve our problem: `listener.incoming()` still blocks the executor.
+Even if connections are handled in separate threads, futures running on the main thread
+are blocked while `listener` waits on incoming connections.
```
Comment (betamos) [inline src/08_example/02_handling_connections_concurrently.md:13]: I'd suggest skipping this section to avoid cognitive load. If you do want to keep it, I'd rephrase it as a caution. Having blocking code on any thread in the runtime is dangerous and runtime-specific. Since async-std supports spawning blocking tasks, it is possible to work around this safely, but that's an executor specific thing that I wouldn't cover right here.

File: src/08_example/04_tests.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# Testing Async Code
```
Comment (betamos) [inline src/08_example/04_tests.md:1]: Sounds pretty general, perhaps we can be more specific that we're testing our code?

File: src/08_example/02_handling_connections_concurrently.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# Handling Connections Concurrently
```
Comment (betamos) [inline src/08_example/02_handling_connections_concurrently.md:1]: The structure here makes sense (concurrency first, then parallelism). But the contents and examples use task spawn, which is already parallel. That's why you have to talk about "the main thread" and so on here, but ideally that could be deferred to the next section.

It would be excellent from an educational perspective if this section was able to simply use a single task with concurrency (e.g. through `for_each_concurrent`), and then you can dedicate the next section to talking about multithreading and parallelism. The code change with adding spawn would be minor, but the effect on the application is big, so it'd be mostly a section talking about how it works.

File: src/08_example/01_running_async_code.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+```
+
+Because we haven't `await`ed or `poll`ed the result of `handle_connection`,
+it'll never run. If you run the server and visit `127.0.0.1:7878` in a browser,
+you'll see that the connection is refused; our server is not handling requests.
+
+We can't `await` or `poll` futures within synchronous code by itself.
+We'll need an executor to handle scheduling and running futures to completion.
+Please consult the section [Choosing an Executor](../404.md) for more information on executors.
+Here, we'll use the `block_on` executor from the `async_std` crate.
```
[the remark this answers is not in the document]
Comment (lbernick) [reply src/08_example/01_running_async_code.md:37]: Changed the terminology here to `runtime`. I think the distinction between a runtime, an executor, and a reactor would be appropriate for the section on choosing an executor.

File: src/08_example/02_handling_connections_concurrently.md (hunk 3 above)
Comment (lbernick) [reply src/08_example/02_handling_connections_concurrently.md:1]: I agree that I'd like to introduce concurrency before talking about multithreading. However, I think the concept of concurrency with only a single task is confusing. Even if we're not blocking the executor while using `for_each_concurrent`, we can still only process one connection at a time. Maybe this could be reframed as a section on avoiding blocking, and then a section on processing connections concurrently by adding parallelism. What do you think?

File: src/08_example/03_combinators.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# Multiple Concurrent Actions Per Request
```
Comment (lbernick) [inline src/08_example/03_combinators.md:1]: @betamos do you think it makes sense to get rid of this section? After addressing your comments I don't think it adds much.

File: src/08_example/02_handling_connections_concurrently.md [hunk 6]
```diff
@@ -1,34 +1,38 @@
...
+First, let's replace our blocking `std::net::TcpListener` with the non-blocking `async_std::net::TcpListener`,
+and update our connection handler to accept an `async_std::net::TcpStream`:
+```rust,ignore
+{{#include ../../examples/08_04_concurrent_tcp_server/src/main.rs:handle_connection}}
+```
 
-This change prevents `listener.incoming()` from blocking the executor
-by allowing us to `await` the next TCP connection on this port.
-Now, the executor can yield to other futures running on the main thread
+This change provides two benefits.
```
Comment (betamos) [inline src/08_example/02_handling_connections_concurrently.md:16]: nit, slight rephrase: Having a stream provides two benefits:

1. It does not block the executor while awaiting the next item.
2. Streams can optionally be processed concurrently, using the `for_each_concurrent` method. Since we need to support multiple simultaneous requests at the same time, we will take advantage of this feature.

Feel free to rephrase however you like. I want to highlight (a) that streams and concurrent execution are different (streams are useful in non-concurrent cases too) and (b) what advantage concurrency brings to our application (handling multiple requests simultaneously).

File: src/08_example/02_handling_connections_concurrently.md [hunk 7]
```diff
@@ -1,34 +1,38 @@
...
 
-Let's update our example to make use of the asynchronous `TcpListener`.
-First, we'll need to update our code to `await` the next incoming connection,
-rather than iterating over `listener.incoming()`:
-```rust
-{{#include ../../examples/08_04_nonblocking_tcp_server/src/main.rs:main_func}}
+The second benefit is that `listener.incoming()` now implements the `Stream` trait,
+allowing us to handle each connection concurrently using its `for_each_concurrent` method.
+We'll need to import the `Stream` trait from the `futures` crate, so our Cargo.toml now looks like this:
+```toml
```
Comment (betamos) [inline src/08_example/02_handling_connections_concurrently.md:24]: Optional: I think there's a "diff" language (replace toml with diff), where you can use `+` before a line to highlight that it's been added. I'm not sure it renders correctly in the book, but this can draw attention to the relevant line.

File: examples/08_04_concurrent_tcp_server/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+use async_std::net::TcpListener;
+use async_std::net::TcpStream;
+use futures::stream::StreamExt;
+
+#[async_std::main]
+async fn main() {
+    let listener = TcpListener::bind("127.0.0.1:7878").await.unwrap();
+    listener
+        .incoming()
+        .for_each_concurrent(/* limit */ None, |stream| async move {
```
Comment (betamos) [inline examples/08_04_concurrent_tcp_server/src/main.rs:11]: nit: Perhaps use `tcp_stream` here in order to avoid confusion with the TcpListener stream that we talk about elsewhere. (Since this is a case of nested streams)

File: src/08_example/04_multithreading.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
+# Adding Parallelism
+Our example so far has largely presented concurrency (using async code)
+as an alternative to parallelism (using threads).
+However, async code and threads are not mutually exclusive.
+Async executors can be either single-threaded or multithreaded.
+For example, the [`async-executor` crate](https://docs.rs/async-executor) used by `async-std`
+has both a single-threaded `LocalExecutor` and a multi-threaded `Executor`.
+
+Tasks do not always need to be run on the thread that created them,
```
Comment (betamos) [inline src/08_example/04_multithreading.md:9]: nit: Neutral value judgment here, something like "tasks can run either on the same thread that created it, or a different thread."

File: src/08_example/04_multithreading.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+and async runtimes often include functionality for spawning tasks onto separate threads.
```
Comment (betamos) [inline src/08_example/04_multithreading.md:10]: nit: often *provide*?

File: examples/08_05_multithreaded_tcp_server/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,26 @@
+// ANCHOR: main_func
+use async_std::net::{TcpListener, TcpStream};
+use async_std::task::spawn;
+
+#[async_std::main]
+async fn main() {
+    let listener = TcpListener::bind("127.0.0.1:7878").await.unwrap();
+
+    loop {
```
Comment (betamos) [inline examples/08_05_multithreaded_tcp_server/src/main.rs:9]: Maybe we could keep using `for_each(spawn(handle_connection))` to keep the code change minimal and still utilize the Stream?

File: src/08_example/05_tests.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
+# Testing the TCP Server
+Let's move on to testing our `handle_connection` function.
+First, we need a `TcpStream` to work with, but we don't want to make a real TCP connection in test code.
```
Comment (betamos) [inline src/08_example/05_tests.md:3]: Hmm, why not make a real connection? Seems like it could be a good integration/e2e test, checking that concurrent connections are possible for instance. No need to add such a test, but to discourage the strategy seems a bit strong.

Perhaps state that here we will write a unit test for the connection handler that will verify the contents of the response.

File: src/08_example/05_tests.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+We could work around this in a few ways.
+One strategy could be to refactor the code to be more modular,
+and only test that the correct responses are returned for the respective inputs.
+
+Another strategy is to connect to `localhost` on port 0.
```
Comment (betamos) [inline src/08_example/05_tests.md:8]: What's wrong with the port you used, 7878?

File: src/08_example/04_multithreading.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+Even if tasks are executed on separate threads, they should still be non-blocking.
+
+Some libraries provide functions for spawning blocking tasks onto separate threads,
+which is useful for running synchronous code from other libraries.
+Tasks are usually required to be `Send`, so they can be moved to separate threads.
+Some libraries also provide functions for spawning non-`Send` tasks onto a thread-local executor.
+Both Tokio and async-std have `task::spawn_blocking` and `task::spawn_local` functions,
+although the async-std versions are unstable.
```
Comment (betamos) [inline src/08_example/04_multithreading.md:18]: nit: Perhaps to avoid future staleness, you could just mention that some runtimes provide methods for this, often called `spawn_blocking` and `spawn_local`.

File: src/08_example/04_multithreading.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
+# Adding Parallelism
```
Comment (betamos) [inline src/08_example/04_multithreading.md:1]: nit: Perhaps more application specific, such as "Serving requests in parallel"?

File: src/08_example/03_combinators.md (hunk 5 above)
Comment (betamos) [reply src/08_example/03_combinators.md:1]: SGTM, given that you cover parallelism in the next section. Regarding the off-tasks of writing to databases etc, I'd like to discuss design patterns for that anyway. If you save the code we can take it separately, which would help this PR merge quicker.

**Review by betamos — CHANGES_REQUESTED** [review]
Thanks again, and sorry for the delayed response.

File: src/08_example/05_tests.md (hunk 13 above)
Comment (lbernick) [reply src/08_example/05_tests.md]: 7878 is fine, but I wrote this due to a preference that the test code should be able to run on any server, where other things might be using port 7878. I believe I also saw this used in the tests for `smol`. No strong preference here.

File: src/SUMMARY.md [hunk 16]
```diff
@@ -26,6 +25,11 @@
   - [`Send` Approximation](07_workarounds/04_send_approximation.md)
   - [Recursion](07_workarounds/05_recursion.md)
   - [`async` in Traits](07_workarounds/06_async_in_traits.md)
+- [Final Project: HTTP Server](08_example/00_intro.md)
+  - [Running Asynchronous Code](08_example/01_running_async_code.md)
+  - [Handling Connections Concurrently](08_example/02_handling_connections_concurrently.md)
+  - [Adding Parallelism](08_example/03_multithreading.md)
```
Comment (betamos) [inline src/SUMMARY.md:31]: I think you forgot to update the title here?

File: src/08_example/01_running_async_code.md [hunk 17]
```diff
@@ -0,0 +1,77 @@
+# Running Asynchronous Code
+An HTTP server should be able to serve multiple clients concurrently.
+Waiting for previous requests to complete before handling the current request can cause performance to suffer.
+The book [solves this problem](https://doc.rust-lang.org/book/ch20-02-multithreaded.html#turning-our-single-threaded-server-into-a-multithreaded-server) by creating a thread pool where each connection is handled on its own thread.
```
Comment (betamos) [inline src/08_example/01_running_async_code.md:4]: Some of these lines exceed the github limit (80 chars?). Perhaps just check if there's an existing limit in the repo and format all lines accordingly? (There's a line break formatting for markdown in VSCode in case you're using that.)

**Review by betamos — APPROVED** [review]
Thanks so much Lee. Great to see this. Just a couple of minor nits left, but LGTM. There are also a couple of spellchecker quirks remaining.

File: src/08_example/01_running_async_code.md (hunk 17 above)
Comment (lbernick) [reply src/08_example/01_running_async_code.md:4]: Looks like the github limit is 125 chars, so I wrapped everything (other than the links which are too long) to be less than that. I didn't see any max line length for this repo.

File: src/08_example/03_multithreading.md [hunk 18]
```diff
@@ -1,26 +1,25 @@
...
 has both a single-threaded `LocalExecutor` and a multi-threaded `Executor`.
 
-Tasks do not always need to be run on the thread that created them,
-and async runtimes often include functionality for spawning tasks onto separate threads.
+Tasks can either be run on the thread that created them or on a separate thread.
+Async runtimes often provide functionality for spawning tasks onto separate threads.
 Even if tasks are executed on separate threads, they should still be non-blocking.
 
-Some libraries provide functions for spawning blocking tasks onto separate threads,
+Some runtimes provide functions for spawning blocking tasks onto separate threads,
```
Comment (tmandry) [inline src/08_example/03_multithreading.md:13]:
```suggestion
Some runtimes provide functions for spawning blocking tasks onto dedicated threads,
```

File: src/08_example/04_tests.md [hunk 19]
```diff
@@ -1,18 +1,20 @@
 # Testing the TCP Server
 Let's move on to testing our `handle_connection` function.
-First, we need a `TcpStream` to work with, but we don't want to make a real TCP connection in test code.
-We could work around this in a few ways.
-One strategy could be to refactor the code to be more modular,
-and only test that the correct responses are returned for the respective inputs.
-
-Another strategy is to connect to `localhost` on port 0.
+First, we need a `TcpStream` to work with.
+In an end-to-end or integration test, we might want a to make a real TCP connection
```
Comment (tmandry) [inline src/08_example/04_tests.md:4]:
```suggestion
In an end-to-end or integration test, we might want to make a real TCP connection
```

File: src/08_example/04_tests.md [hunk 20]
```diff
@@ -1,18 +1,20 @@
 # Testing the TCP Server
 Let's move on to testing our `handle_connection` function.
-First, we need a `TcpStream` to work with, but we don't want to make a real TCP connection in test code.
-We could work around this in a few ways.
-One strategy could be to refactor the code to be more modular,
-and only test that the correct responses are returned for the respective inputs.
-
-Another strategy is to connect to `localhost` on port 0.
+First, we need a `TcpStream` to work with.
```
Comment (tmandry) [inline src/08_example/04_tests.md:3]: I'd suggest breaking this paragraph up

```suggestion

First, we need a `TcpStream` to work with.
```

File: src/08_example/04_tests.md [hunk 21]
```diff
@@ -1,18 +1,20 @@
...
+In an end-to-end or integration test, we might want a to make a real TCP connection
+to test our code.
+One strategy for doing this could be to connect to `localhost` on port 0.
 Port 0 isn't a valid UNIX port, but it'll work for testing.
 The operating system will return a connection on any open TCP port.
```
Comment (tmandry) [inline src/08_example/04_tests.md]: I don't think the pre-existing text seems quite right here, what about this?

```suggestion
One strategy for doing this is to start a listener on `localhost` port 0.
Port 0 isn't a valid UNIX port, but it'll work for testing.
The operating system will pick an open TCP port for us.
```

**Review by tmandry — APPROVED** [review]
Looks good! I left a few more nits and suggestions, otherwise I think this is ready to merge.

**Comment by tmandry** [comment]
Merged. Thank you!
